### PR TITLE
feat: expose cache to custom functions

### PR DESCRIPTION
**Pull request by P0lip**
**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

I hope I got the documentation right :sweat_smile:

### CODE REVIEW

File: docs/guides/custom-functions.md [hunk 1]
```diff
@@ -233,6 +233,18 @@ module.exports = (targetVal, _opts, paths) => {
 };
 ```
 
+Furthermore, if your function performs any fs calls to obtain dictionaries and similar, you may want to leverage cache.
```
Comment (nulltoken) [inline docs/guides/custom-functions.md:236]: @P0lip The doc is pretty clear to me. However, from a user standpoint, there's no information with regards to _how_ I can leverage the cache. 

Could you please add a simple sample?

File: docs/guides/custom-functions.md [hunk 2]
```diff
@@ -233,6 +233,18 @@ module.exports = (targetVal, _opts, paths) => {
...
+Each custom function is provided with its **own** cache instance that has a function-live lifespan.
+What does "function-live lifespan" mean? It means the cache is persisted for the whole life of a particular function.
+The cache will be retained between subsequent function calls and is never invalidated unless you compile the function again, i.e. load a ruleset again.
+In other words:
+- if you Spectral programmatically via JS API and your ruleset remains unchanged, all subsequent `spectral.run` calls will invoke custom functions with the same cache instance.
+As soon as you set a ruleset using `setRuleset` or `loadRuleset` method, each custom function will receive a new cache instance.
+- if you are a CLI user, the cache will never be invalidated.
```
Comment (nulltoken) [inline docs/guides/custom-functions.md:243]: @P0lip within the timespan of a CLI run or even across runs?

File: docs/guides/custom-functions.md [hunk 3]
```diff
@@ -233,6 +233,18 @@ module.exports = (targetVal, _opts, paths) => {
...
+
+Please bear in mind that while you are welcome to store certain kind of data, using cache for exchanging information between subsequent function calls it strongly discouraged.
```
Comment (nulltoken) [inline docs/guides/custom-functions.md:245]: @P0lip From a security standpoint, it's also highly discouraged to store sensitive information in this cache.

File: src/__tests__/linter.jest.test.ts [hunk 4]
```diff
@@ -66,6 +70,48 @@ describe('Linter', () => {
     );
   });
 
+  it('should expose function-live lifespan cache to custom functions', async () => {
```
Comment (nulltoken) [inline src/__tests__/linter.jest.test.ts:73]: @P0lip Could you please add a test showing that the cache is not shared among different functions?

File: docs/guides/custom-functions.md (hunk 2 above)
Comment (P0lip) [reply docs/guides/custom-functions.md:243]: Just a timespan of a CLI. Will include it in the description.

File: docs/guides/custom-functions.md (hunk 1 above)
Comment (P0lip) [reply docs/guides/custom-functions.md:236]: Ah yeah, forgot about an example.

**Pull request merged**